engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/blueprint/NadelExecutionBlueprintFactory.kt

```diff
@@ -567,3 +599,138 @@ private class SharedTypesAnalysis(
         }
     }
 }
+
+private class TypeOwnership(
+    private val overallSchema: GraphQLSchema,
+    private val services: List<Service>,
+    private val typeRenames: List<NadelTypeRenameInstruction>,
+) {
+    fun getTypesToServicesThatDefineThem(): Map<String, Set<Service>> {
+        return getTypesThatServiceDefines()
+            .entries
+            .asSequence()
+            .flatMap { (service, typesDefined) ->
+                typesDefined
+                    .asSequence()
+                    .map {
+                        it to service
+                    }
+            }
+            .groupBy {
+                it.first
+            }
+            .mapValues { it: Map.Entry<String, List<Pair<String, Service>>> ->
+                it.value // Get the List<Pair<…>> out
+                    .asSequence()
+                    .map {
+                        it.second // Get the Service out of the Pair<String, Service>
+                    }
+                    .toSet()
+            }
+    }
+
+    fun getTypesThatServiceDefines(): Map<Service, Set<String>> {
```

Comment:
Perhaps a better name is types that a service exposes. Because of shared types, we can't rely on `service.definitionRegistry`, see commentary below for more.